accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,671 @@
+    - Enumerations
+  - Reference types
+    - **Pointer types**
+    - Interface types
+    - Self-describing types
+      - Arrays
+      - Class types
+        - User-defined classes
+        - Boxed value types
+        - Delegates
+
+## Overidden members for a modified type
+Below are the non-trivial behaviors overridden by a modified type.
+
+### `bool IsModifiedType { get; }`
+Returns `true`.
+
+### `Type UnderlyingSystemType { get; }`
+The underlying type, such as `typeof(Int32)`.
+
+### `ToString()`
+Returns `UnderlyingSystemType.ToString()`. We could return a string representation of the underlying type plus its custom modifiers, but that is not considered a high enough priority at this time and the grammar would be somewhat arbitrary.
+
+### `Name`, `FullName`, `AssemblyQualifiedName`, `Assembly` and `Namespace` properties
+A modified type is an internal class called "System.RuntimeModifiedType" class that derives from a "System.ModifiedType" which derives from "System.Type". So these methods just return the expected values there.
+
+### `GetFunctionPointerParameters()`
+This returns modified types for each function pointer parameter which allows the user to call `GetRequiredCustomModifiers()` or `GetOptionalCustomModifiers()`.
+
+This is listed here since this method also works on unmodified types, although unmodified types don't allow `GetRequiredCustomModifiers()` or `GetOptionalCustomModifiers()` to be called (an `InvalidOperationException` is thrown).
+
+# Proposed APIs
+## System.Type
+```diff
+namespace System
+{
+    public abstract class Type
+    {
+        // Throws InvalidOperationException if IsFunctionPointer = false.
++       public virtual Type GetFunctionPointerReturnType();
++       public virtual Type[] GetFunctionPointerParameterTypes();
++       public virtual bool IsFunctionPointer { get; }
+
++       public virtual bool IsUnmanagedFunctionPointer { get; }
+
+        // Returns whether the type is a modified type.
+        // A modified type is either:
+        //  - A function pointer obtained though FieldInfo\PropertyInfo\ParametersInfo.GetModified*Type().
+        //  - A function pointer parameter obtained by 'GetFunctionPointerParameterTypes()[index]' from a modified function pointer type.
++       public virtual bool IsModifiedType { get; }
+
+        // Updated semantics: if IsModifiedType = true, returns the unmodified type such as `typeof(Int32)`.
+        public virtual bool UnderlyingSystemType { get; }
+
+        // These only work if IsModifiedType = true. InvalidOperationException otherwise.
++       public virtual Type[] GetRequiredCustomModifiers();
++       public virtual Type[] GetOptionalCustomModifiers();
+        // Throws InvalidOperationException if UnderlyingSystemType.IsFunctionPointer = false
++       public virtual Type[] GetFunctionPointerCallingConventions(); 
+
+        // This is used with 'Type.GetMethod(..., Type[] types)' to find a method with parameters that
```

Comment:
You can overload based on calling convention. For example, the following compiles fine:
```
static void foo(delegate* unmanaged[Stdcall]<void> f) { }
static void foo(delegate* unmanaged[Cdecl]<void> f) { }
```
Is this going to find the first overload at random? Or are going to allow passing modified types to `GetMethod` calls?

Reply:
The right thing to do is to throw `AmbiguousMatchException` and then, I suppose, add a `MakeModifiedFunctionPointer()` overload that takes the extra custom mods.

Adding such an overload would be ugly (2 dimension jagged array?) unless we add a helper class, like the "MethodSignature" proposed in the design.

After thinking about this, I may move this feature to the "future" section. For now, to avoid a breaking change later, we throw `NotSupportedException` if there is a function pointer type in the `Type[]` passed to `Type.GetMethod()`. The workaround is the user would need to call `GetMethods()` and then manually loop through them and inspect the custom modifiers on the parameters themselves.

Reply:
(design updated for this)